src/main/java/homer/model/temperaturechangers/AbstractTemperatureChanger.java

```diff
@@ -0,0 +1,107 @@
+package homer.model.temperaturechangers;
+
+import homer.api.DeviceInfo;
+import homer.common.Temperature;
+import homer.common.TemperatureFactory;
+import homer.model.environment.Environment;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+
+/**
+ * Uses template method on updateTick. At a default intensity, it changes the temperature by 
+ * 1 celsius degree each hour;
+ */
+public abstract class AbstractTemperatureChanger implements TemperatureChanger {
+
+    private static final double NORMALIZER = 1 / (1000d * 3600); // used to normalize the intensity
+    private final double maxIntensity;
+    private final double minIntensity;
+    private double intensity;
+    private final Environment environment;
+    private final DeviceInfo info;
+    private Temperature minTemperature = TemperatureFactory.fromKelvin(0);
+    private Temperature maxTemperature = TemperatureFactory.fromCelsius(1000);
+
+    /**
+     * @param minIntensity
+     * @param maxIntensity
+     * @param environment
+     * @param info
+     */
+    @SuppressFBWarnings(value = "EI_EXPOSE_REP", 
+        justification = "Updating a reference is better than reallocating objects on the heap")
+    public AbstractTemperatureChanger(final double minIntensity, final double maxIntensity, 
+        final Environment environment, final DeviceInfo info)  {
+        this.info = info;
+        this.minIntensity = minIntensity;
+        this.maxIntensity = maxIntensity;
+        this.intensity = minIntensity;
+        this.environment = environment;
+    }
+
+    @Override
+    public final DeviceInfo getInfo() {
+        return this.info;
+    }
+
+    @Override
+    public final Double getState() {
+        return this.intensity;
+    }
+
+    @Override
+    public final Double getMinValue() {
+        return this.minIntensity;
+    }
+
+    @Override
+    public final Double getMaxValue() {
+        return this.maxIntensity;
+    }
+
+    @Override
+    public final void setState(final Double value) {
+        this.intensity = Math.max(Math.min(maxIntensity, value), minIntensity);
+    }
+
+    /**
+     * @return maximum temperature allowed
+     */
+    public final Temperature getMaxTemperature() {
+        return this.maxTemperature;
+    }
+
+    @Override
+    public final void setMaxTemperature(final Temperature temperature) {
+        this.maxTemperature = temperature;
+    }
+
+    /**
+     * @return minimum temperature allowed
+     */
+    public final Temperature getMinTemperature() {
+        return this.minTemperature;
+    }
+
+    @Override
+    public final void setMinTemperature(final Temperature temperature) {
+        this.minTemperature = temperature;
+    }
+
+    /**
+     * 
+     * @return the environment associated with the object 
+     */
+    protected final Environment getEnvironment() {
+        return this.environment;
+    }
+
+    /**
+     * 
+     * @return normalization value
+     */
+    public double getNormalizer() {
```

Comment:
Seems a bit odd, is it of any use?

Reply:
Yeah, it's to shut the linter up. i could also mark normalizer as static and final, and i actually think it would be cleaner

Reply:
I don't understand what the linter was complaining about.
If this is only for internal use it should not be public.
I did get confused at first, because it made me think it was returning the normalized value of the intensity (normalized means returning a value between 0.0 and 1.0)

Reply:
Yeah, honestly i thought it was a good variable name, but maybe `scaler` fits better?

Reply:
Do you have the error message?
What if you use a protected getter for the private field?

Reply:
>What if you use a protected getter for the private field?

No errors. I'll just change this, rename `normalizer` and merge if that's fine with you

Reply:
Ok!